clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -156,10 +160,13 @@ public FilterResult filterTo(LogEntryFilter filter, ByteBuffer buffer) {
                     maxTimestamp = info.maxTimestamp;
                     shallowOffsetOfMaxTimestamp = info.shallowOffsetOfMaxTimestamp;
                 }
+                if (shallowEntry.offset() > maxOffset) {
```

Comment:
Hmm, we can't use shallowEntry.offset() since not all messages will be retained. Perhaps we could just maintain maxOffset as we add messages to retainedEntries.